<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script type="text/javascript">
        //异步
        //setTimeout
        // console.log("第一行");
        // var timer1 = setTimeout(() => {
        //     console.log("1秒后执行");
        // }, 1000);//1秒后才打印结果
        // console.log("第二行");

        // setTimeout(() => {
        //     clearTimeout(timer1);//中断timer1执行
        // }, 500); //500毫秒后中断timer1的执行，这样就不输出“1秒后执行”了

        //setInterval
        // var timer2 = setInterval(() => {
        //     let date = new Date();
        //     console.log(`${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}`);
        // }, 1000);

        // setTimeout(() => {
        //     clearInterval(timer2);
        // }, 5000);//5秒后取消定时

        //异步 创建promise
        // var promise = new Promise(resolve => {
        //     setTimeout(() => {
        //         resolve("执行成功");
        //     }, 2000);
        // })

        // promise.then(value => console.log(value));
        // console.log("在promise之前执行");

        // //promise 异常捕获
        // var promise2 = new Promise((resolve, reject) => {
        //     setTimeout(() => {
        //         reject("执行失败");
        //     }, 2000);
        // });
        // promise2.catch(error => { console.log(error) });//捕获promise异常信息

        //promise 链式调用
        // new Promise((resolve, reject) => {
        //     setTimeout(() => {
        //         resolve(1);//传递结果1   resolve=1
        //     }, 1000);
        // }).then(value => {
        //     console.log(value);
        //     return value+10;//输出1，并返回1+10
        // }).then(value => {
        //     console.log(value);
        //     return new Promise(resolve => resolve(value+20));//输出11，并返回新的promise（resolve=11+20）
        // }).then(value => {
        //     console.log(value);
        // });

        // new Promise((resolve, reject) => {
        //     setTimeout(() => {
        //         reject("promise失败");//传递结果1   resolve=1
        //     }, 1000);
        // }).then(value => {
        //     console.log(value);
        //     return value+10;//输出1，并返回1+10
        // }).then(value => {
        //     console.log(value);
        //     return new Promise(resolve => resolve(value+20));//输出11，并返回新的promise（resolve=11+20）
        // }).then(value => {
        //     console.log(value);
        // }).catch(error => {//第一个环节异常后，中间环节就不执行了，直接跳转到catch
        //     console.log(error);
        // });

        //promise   多个promise同时执行
        var p1 = new Promise(resolve => {
            setTimeout(()=>{
                resolve(1);
            },1000);
        });
        var p2 = new Promise(resolve => {
            setTimeout(()=>{
                resolve(2);
            },2000);
        });
        var p3 = new Promise(resolve => {
            setTimeout(()=>{
                resolve(3);
            },500);
        });
        //promise返回结果的顺序和all定义的数组一致，与返回时间的先后无关
        // Promise.all([p1,p2,p3]).then(values => console.log(values));//[1, 2, 3]

        //异步 async & await
        //async标识声明异步函数   1秒后执行
        async function async1(){
            
            //await必须在带async关键字的函数中调用，用户返回async函数的结果
            //await有阻塞的效果，await语句不执行的话，后续语句无法执行
            let result = await async2();
            console.log("async2返回结果："+result);

            setTimeout(()=>{
                console.log("async1 执行完毕");
            },1000);

            try {
                let result2 = await async3();
            } catch (error) {
                console.log(error);
            }
        }
        async1();//调用异步执行的函数
        // console.log(async1());//返回结果还是一个promise
        console.log("主程序执行");

        async function async2(){
            return new Promise(resole => {
                setTimeout(() => {
                    resole(10);//1秒后返回结果10
                },1000);
            });
        }

        async function async3(){
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                    reject("async3执行异常");
                },1000);
            });
        }

    </script>
</head>

<body>

</body>

</html>